### ISSUE TITLE
Continue processing events even if one callback dies

### ISSUE THREAD

**Issue by preaction** [preaction/Beam-Emitter#9]
It'd be nice to be able to safely call all the callbacks, even if one of the callbacks throws an exception.

**Comment by kentfredric** [comment]
I think this is the sort of thing the person emitting the event should be in control of.

so maybe a callback for handling failure callback? idk.

I think this is starting to approach Promises.

**Comment by preaction** [comment]
So, `emit_safe` and `emit_safe_args` perhaps? And then, yeah, the emitter has an event `emit_error` or `event_error` or just `error` that gets called?

**Comment by kentfredric** [comment]
Oh, you were meaning to have the error handler centralised on the object?

I was thinking more a coderef that's per-event for some reason.

**Comment by kentfredric** [comment]
like  ->emit(  'eventname', { class => "Foo::Bar", error => sub {  }} , .... )

**Comment by preaction** [comment]
Yeah, that looks nicer. This way, if they want to emit an event, they can just do exactly:

```
$self->emit( 'event', error => $self->curry::weak::emit( 'error' ) );
```

Which is explicit rather than implicit and keeps us from prematurely increasing the complexity of Beam::Emitter, which I like. If that pattern becomes common, we can add a default error handler that uses the "emitter_error" event or something (though we're in danger of namespace collision).

`emit_args` doesn't get the same benefit though, since we can't assume anything about the args passed in. That'd need an `emit_args_safe` version. Or for now we could just say "meh, your problem for using emit_args", which I'm totally fine with (emit_args was added after a coworker decided to write their own event emitter without telling me or asking for the change they wanted because that's how terrible decisions are always made).

Indeed with your way we can still have the `_safe` versions for consistency between `emit` and `emit_args`, but have them simply ignore errors. Since that is an incredibly bad idea, and we will yell and gnash about that in the docs, hopefully they'll use this version instead.

**Comment by kentfredric** [comment]
Though I'm kinda starting to despair about the growth of methods on the emitter classes after role composition.

Like, when is it ever useful to do $not_me->emit() ?

**Comment by preaction** [comment]
Yeah. It's not as though we _must_ do all of this right away. The initial add is just the error handler. But then if anyone asks, we have a plan for the rest of it.

**Comment by preaction** [comment]
So, after some thought, it's not as simple as just providing an error callback and calling it:

* What arguments does the error callback get? I'm thinking it needs to be:
    1) The original event
    2) The listener that died
    3) The exception thrown / error message
    It would probably be best if this is all encapsulated into an object like Beam::Event::Error
* What does one need to be able to debug a bad listener? I'm guessing they need a stack trace from the exception, but also a stack trace from when the offending listener was created. This means adding a stack trace to the Beam::Listener object when it's created. We must be careful to not make it harder to debug by hiding / messing with the stack traces.
* Should error handlers be given a choice to continue with the exception propagation or stop it? I think this question is the easiest to answer: Since it's the Emitter itself deciding, it's easy enough to say "if there's an error callback, there's not an exception", but I'd like to avoid having a bit of code that says `if ( $args{error} ) { eval { &do_thing } } else { &do_thing }`... That, and there's still `$event->prevent_default`.
